GreptimeTeam/greptimedb
# test: sqlness upgrade compatibility tests

I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

## What's changed and what's your intention?

Add a way to test inter-version compatibility in sqlness

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- add a `-- SQLNESS ARG version=v0.x.x` which will restart greptime from given binary in a directory
- add a simple poc sqlness test, to run those tests
- add ability to pull different version binary from github release when encounter `version=xxx`, default to pull, but can set to not pull and manually download it. 
- more sqlness tests can be added later in separate PR to prevent bloat
- add ability to setup a standalone etcd for metasrv metadata storage so restarting metasrv wouldn't clean up metadata

to playaround with this test, run:
 ```bash
cargo sqlness -c $(pwd)/tests/upgrade-compat --store-addrs "0.0.0.0:2379" --setup-etcd 
``` 
(make sure to use absolute path for case dir, the `-c` option, can also add `--preserve-state` to look into `/tmp/sqlness*****/greptime-sqlness-standalone.log` to make sure it actually switched version, note this require docker and create a `etcd` docker container for metadata storage during testing)

## Summary
- How does this PR work? Need a brief introduction for the changed logic (optional)
Basically just restart greptime using another binary
- Describe any limitations of the current code (optional)
Well this can't currently test http api, but I intended to add ability to invoke bash in sqlness tests,
so one can just call `curl` in sqlness tests
## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
